_drafts/choosing_or_building_an_embedded_db.md

```diff
@@ -0,0 +1,330 @@
+
+## Why You Should Care About NVM
+
+Nearly every embedded system needs to persist data across power cycle.
+A smart fishtank needs to remember the configured salinity and oxygen setpoints,
+an e-bike may store a geofence and riding log,
+and a smart home hub must store device, group, and schedule information.
+
+Let's assume for now that we're talking about a small embedded system,
+like an STM32 with only the built-in flash available for persistant storage.
+Is there an obvious solution? What happens if we just write our data to flash
+and read it back on reboot?
+
+For the fishtank example, the data we want to persist could be organized into a single
+struct which is read from a specific flash address on boot and written to that address
+when the user changes the settings:
+
+```c
+typedef struct {
+    int salinity;
+    int oxygen;
+} fishtank_config_t;
+
+fishtank_config_t config;
+
+void load() {
+  flash_read(CONFIG_ADDR, &config, sizeof(config));
+}
+
+void save() {
+  flash_erase_page(CONFIG_ADDR);
+  flash_program(CONFIG_ADDR, &config, sizeof(config));
+}
+```
+
+This solution may "just work" for an MVP fishtank,
+but as the product evolves we are going to run into a few issues:
+
+ 1. The persisted data is not portable due to endianness and integer size differences,
+ 2. The config will be corrupted if power is lost during the `save()` function,
+ 3. If a `temperature` field is added to the struct, it will be read as `-1` upon 
+    firmare upgrade, which may be unexpected.
+
+So we see that even for the simplest products, persisting data to flash
+presents some mantainability challenges.
+
+## Aside: Portability of Records
+
+Thinking ahead, you may eventually wish to run your fishtank code on 
+different hardware either because of an upgrade to the product,
+a new version of the product, or because you want to be able to load
+up records from the embedded system inside a mobile app or a server.
+
+That new processor may be 64-bit processor, or may even use a 
+different endianness than the microcontroller. Taking a closer look at
+the `fishtank_config_t` struct, we see that the basic `int` type is
+used, which uses 8 bytes on 64-bit versus 4 bytes on 32-bit,
+and the endianness (byte order) depends on the processor.
+
+### Integer Size and Alignment
+
+With the definition of `fishtank_config_t` above,
+`sizeof(fishtank_config_t)` will be different for different
+platforms. Compiling on a 64-bit MacBook where Apple
+kept `int` as 32-bit with 32-bit alignment, you'll get the 
+same 8-byte structure as on a 32-bit microcontroller.
+However, if you compile on an AWS EC2 instance
+(as you may want to do if there is a cloud component to 
+your fishtank), then you get a 16-byte structure because
+`int` there is 8 bytes.
+
+The bottom line for portability is, to use the definitions in `stdint.h`
+and instruct the compiler to pack your structure to remove any padding
+which would otherwise be added to encourage alignment to word boundries:
```

Comment:
```suggestion
For portability, use the fixed-width definitions in `stdint.h`
and instruct the compiler to pack your structure to remove any padding
which would otherwise be added to encourage alignment to word boundries:
```